org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/services/GenericEndpoint.java

```diff
@@ -63,7 +63,13 @@ protected void recursiveFindRpcMethods(Object current, Set<Class<?>> visited, Se
 					Method method = methodInfo.method;
 					Object[] arguments = this.getArguments(method, arg);
 					return (CompletableFuture<Object>) method.invoke(current, arguments);
-				} catch (InvocationTargetException | IllegalAccessException e) {
+				} catch (InvocationTargetException e) {
+					if (e.getTargetException() instanceof ResponseErrorException) {
```

Comment:
Looks good, just some notes:

This could be a breaking change for existing exception handlers.

If the behavior is changed we could even throw all target exceptions which are an instance of `RuntimeException` directly, without wrapping?
E.g. 

```
if (e.getTargetException() instanceof RuntimeException) {
  throw (RuntimeException) e.getTargetException();
}
```
And may be even remove always the `InvocationTargetException` which is a technical detail because reflection is used?
```
if (e.getTargetException() instanceof RuntimeException) {
  throw (RuntimeException) e.getTargetException();
}
throw new RuntimeException(e.getTargetException());
```

Reply:
@ivy-cst Just wondering, why do you think this could be a breaking change for existing exception handlers? I assume that any exception handlers should have been prepared to handle a `ResponseErrorException` thrown directly, in the same way as `DEFAULT_EXCEPTION_HANDLER` does.

In general, I think that it would have been better to use a `CompletionException` as a wrapper rather than a `RuntimeException` here, or perhaps even ["sneaky throws"](https://www.baeldung.com/java-sneaky-throws) without any wrapping at all.

Reply:
I'm not worried about this being a breaking change - I don't think that we document what is expected beyond what `DEFAULT_EXCEPTION_HANDLER` does already and the current `DEFAULT_EXCEPTION_HANDLER` can handle these new cases.

I do agree that wrapping in a `CompletionException` and discarding the `InvocationTargetException` is a reasonable thing to do.

I see that `DEFAULT_EXCEPTION_HANDLER` has a bug itself regarding `InvocationTargetException`. It has this code:

```java
if ((throwable instanceof CompletionException || throwable instanceof InvocationTargetException)
				&& throwable.getCause() instanceof ResponseErrorException)
```

but in the case of InvocationTargetException this will never be true because InvocationTargetException's cause is always null, and thus never an instanceof `ResponseErrorException`. 

So in conclusion more updates are needed. I'll revisit soon, but if someone want to take this up in the meantime that would be great because my allocated LSP4J time is running out quick.

Reply:
> I have implemented an alternative approach in a separate commit, so it can easily be reviewed and reverted if necessary.

I'll have a look.

> The Javadoc for InvocationTargetException.getTargetException() says that getCause() is now the preferred means of obtaining the target exception, so this code looks OK.

Sorry - that was misinformation on my part. Thanks for checking and sorry for the noise.

Reply:
Lets go with wrapping in CompletionException as that looks good. I will squash your change when I next push this PR.